### DISCUSSION TITLE
Neorg on NixOS

### DISCUSSION THREAD

**Discussion by florentc** [nvim-neorg/neorg#1392] [Q&A]
I am looking for a way to setup the latest version of Neorg declaratively on NixOS in a home-manager Neovim configuration.

### The attempt (one among countless previous ones)

Fetched the latest release of neorg from GitHub (the current one on nixpkgs still has a bug when deleting the last line of a file)
```nix
let neorg_custom = pkgs.vimUtils.buildVimPlugin {
  pname = "neorg";
  version = "8.4.1";
  src = pkgs.fetchFromGitHub {
    owner = "nvim-neorg";
    repo = "neorg";
    rev = "v8.4.1";
    sha256 = "5jgCJZMG5V+2U/GWmaNWmdsEzIwLo+8hKE88YQZqHUU=";
  };
};
in
```

I then use it in my neovim home-manager module, and declare the dependencies explicitly (on nixpkgs, these dependencies are not provided either anyway) :

```nix
    programs.neovim = {
      enable = true;
      defaultEditor = true;
      plugins = with pkgs.vimPlugins; [
        neorg_custom nvim-nio nui-nvim plenary-nvim
        neorg-telescope
        nvim-cmp
        nvim-treesitter.withAllGrammars
      ];
      extraLuaPackages = luaPkgs : [
        luaPkgs.pathlib-nvim # For neorg
        luaPkgs.lua-utils-nvim # For neorg
      ];
    };
```

### The problem

The `norg_meta` parser is not provided by `nvim-treesitter.withAllGrammar`. Thus, Neorg sometimes fail (e.g. when opening journal toc).

Since treesitter is managed by Nix, `:Neorg sync-parsers` fails because it tries to write in the Nix store.

The `norg` parser works (is it part of the treesitter distribution on nixpkgs maybe? is that version up to date anyway?).

### The question

Could somebody help me find a decent way to get Neorg up and running on NixOS in a declarative way?

**Comment by alex-demetriou** [comment]
The norg meta grammar is in `unstable`, so it should make its way into `24.05`.

If you don't want to move to unstable completely, you can do something like this in your config.

```nix
{
  plugins =
    let
      unstable = import (
        builtins.fetchGit {
          url = "https://github.com/nixos/nixpkgs/";
          ref = "refs/heads/nixos-unstable";
        }
      );
    in
    [
      # your other plugins
      unstable.vimPlugins.nvim-treesitter.withAllGrammars
    ];
}

```


Ref https://github.com/NixOS/nixpkgs/pull/270304

**Reply by florentc** [reply]
The norg-meta grammar is indeed in nixpkgs (`tree-sitter-grammars.tree-sitter-norg-meta`) but it is [not a grammar that is part of `nvim-treesitter.withAllGrammars`](https://github.com/NixOS/nixpkgs/blob/5c95d720282c70aae65447fe21c380a319010d17/pkgs/applications/editors/vim/plugins/nvim-treesitter/generated.nix#L1665).

**Comment by Rexcrazy804** [comment]
I have this in my configuration
```nix
{
  treesitter = let
    nts = pkgs.vimPlugins.nvim-treesitter;
    tsg = pkgs.tree-sitter-grammars;
    norgG = [tsg.tree-sitter-norg tsg.tree-sitter-norg-meta];
  in
    nts.withPlugins (_: nts.allGrammars ++ norgG);
}
```
